File: src/main/java/ru/darkkeks/vkmirror/autoreg/RegisterAction.java
package ru.darkkeks.vkmirror.autoreg;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

class RegisterAction extends BotFatherAction<String> {

    private static final Logger logger = LoggerFactory.getLogger(RegisterAction.class);

    private final Pattern tokenPattern = Pattern.compile("\\d+:\\w+");
    private final Pattern alreadyTaken = Pattern.compile("already taken");

    public static final String NEW_BOT = "/newbot";

    private String username;
    private String name;

    public RegisterAction(BotAutoReg autoReg, String username, String name) {
        super(autoReg);
        this.username = username;
        this.name = name;
    }

    @Override
    public void execute() {
        String newBot = getAutoReg().executeCommand(NEW_BOT);

        int timeout = getAutoReg().getTimeoutSeconds(newBot);
        if (timeout != BotAutoReg.NO_TIMEOUT) {
            getCompletableFuture().completeExceptionally(new BotFatherTimeoutException(timeout));
            return;
        }

        String afterName = getAutoReg().executeCommand(name);
        String afterUsername = getAutoReg().executeCommand(username);

        if (alreadyTaken.matcher(afterUsername).find()) {
            getCompletableFuture().completeExceptionally(
                    new IllegalStateException(String.format("Bot username %s is already taken", username)));
            return;
        }

        Matcher matcher = tokenPattern.matcher(afterUsername);
        boolean foundToken = matcher.find();
        if (foundToken) {
            String token = matcher.group(0);
            finish(token);
        } else {
            logger.error("No token in BotFather response!:\n{}", afterUsername);
            getCompletableFuture().completeExceptionally(new IllegalStateException("No token in response"));
        }
    }
}

File: src/main/java/ru/darkkeks/vkmirror/bot/RegisterAction.kt
package ru.darkkeks.vkmirror.bot

import ru.darkkeks.vkmirror.util.createLogger

val ALREADY_TAKEN = "already taken".toRegex()
val TOKEN = """\d+:\w+""".toRegex()

class RegisterAction(autoReg: BotAutoReg,
                     val username: String,
                     val name: String) : BotFatherAction<String>(autoReg) {

    override suspend fun execute() {
        val newBot = autoReg.executeCommand(NEW_BOT)

        val timeout = getTimeout(newBot)
        if (timeout != null) {
            callback.completeExceptionally(BotFatherTimeoutException(timeout))
            return
        }

        val afterName = autoReg.executeCommand(name)
        val afterUsername = autoReg.executeCommand(username)

        if (ALREADY_TAKEN.find(afterUsername) != null) {
            callback.completeExceptionally(IllegalStateException("Bot username $username is already taken"))
            return
        }

        val match = TOKEN.find(afterUsername)
        if (match != null) {
            callback.complete(match.value)
        } else {
            logger.error("No token in BotFather response!:\n$afterUsername")
            callback.completeExceptionally(IllegalStateException("No token in response"))
        }
    }

    companion object {
        val logger = createLogger<RegisterAction>()
    }
}